byterun/array.c

```diff
@@ -241,11 +366,15 @@ CAMLprim value caml_make_array(value init)
         res = caml_check_urgent_gc(res);
       }
       for (i = 0; i < size; i++) {
-        Store_double_field(res, i, Double_val(Field(init, i)));
+        double d = Double_val(Field(init, i));
+        Store_double_flat_field(res, i, d);
       }
       CAMLreturn (res);
     }
   }
+#else
+  return init;
```

Comment:
This might be worth a comment explicitly noting that a copy is not required, since the name of the function makes it sound like it should always copy.  Or maybe you should change the name of the function, in fact.

Reply:
(This has been addressed)